stdlib/string.ml

```diff
@@ -280,3 +280,100 @@ let get_int32_le s i = B.get_int32_le (bos s) i
 let get_int32_be s i = B.get_int32_be (bos s) i
 let get_int64_le s i = B.get_int64_le (bos s) i
 let get_int64_be s i = B.get_int64_be (bos s) i
+
+(* Spellchecking *)
+
+let uchar_utf_8_byte_decode_length = function
+  | '\x00' .. '\x7F' -> 1
+  | '\x80' .. '\xC1' -> 0
+  | '\xC2' .. '\xDF' -> 2
+  | '\xE0' .. '\xEF' -> 3
+  | '\xF0' .. '\xF4' -> 4
+  | _ -> 0
+
+let utf_8_uchar_length s =
+  let slen = length s in
+  let i = ref 0 and ulen = ref 0 in
+  while (!i < slen) do
+    let dec_len = uchar_utf_8_byte_decode_length (unsafe_get s !i) in
+    i := (!i + if dec_len = 0 then 1 (* count one Uchar.rep *) else dec_len);
+    incr ulen;
+  done;
+  !ulen
+
+let uchar_array_of_utf_8_string s =
+  let slen = length s in (* is an upper bound on Uchar.t count *)
+  let uchars = Array.make slen Uchar.max in
+  let k = ref 0 and i = ref 0 in
+  while (!i < slen) do
+    let dec = get_utf_8_uchar s !i in
+    i := !i + Uchar.utf_decode_length dec;
+    uchars.(!k) <- Uchar.utf_decode_uchar dec;
+    incr k;
+  done;
+  uchars, !k
+
+let edit_distance' ?(limit = Int.max_int) s (s0, len0) s1 =
+  if limit <= 1 then (if equal s s1 then 0 else limit) else
+  let[@inline] minimum a b c = Int.min a (Int.min b c) in
+  let s1, len1 = uchar_array_of_utf_8_string s1 in
+  let limit = Int.min (Int.max len0 len1) limit in
+  if Int.abs (len1 - len0) >= limit then limit else
+  let s0, s1 = if len0 > len1 then s0, s1 else s1, s0 in
+  let len0, len1 = if len0 > len1 then len0, len1 else len1, len0 in
+  let rec loop row_minus2 row_minus1 row i len0 limit s0 s1 =
+    if i > len0 then row_minus1.(Array.length row_minus1 - 1) else
+    let len1 = Array.length row - 1 in
+    let row_min = ref Int.max_int in
+    row.(0) <- i;
+    for j = Int.max 1 (i - limit - 1) to Int.min len1 (i + limit + 1) do
```

Comment:
There is an overflow issue with `js_of_ocaml` where  `Sys.max_string_length` is greater than `Int.max_size / 2`. However, it sounds sensible to raise an error in this case.

Reply:
Are you suggesting something should be done ? That would show up if you start computing the (unlimited) edit distance between ~1GB strings without a `limit`. Not sure if that use case makes much sense.

Reply:
Good question, I was thinking of raising `Invalid_argument "edit_distance does not support string larger than Int.max_int /2"` to avoid computing a wrong value in this corner case, but this puts quite a strain on usability to warn against a very narrow corner case.

Reply:
We could also also perhaps detect the overflow and replace with `len1`. It seems that in `js_of_ocaml` `Sys.max_string_length + Sys.max_string_length < 0`.

That is replace  

```
Int.min len1 (i + limit - 1)
```
by 
```
let max = 
  let max = Int.min len1 (i + limit - 1)
  if max < 0 then len1 else max 
```
But it's a bit more comparisons in the loop.

Reply:
So I went with that there's no measurable effect.